Add a grabber for SICK 2D LiDAR: tim

ysuzuki19 opened the pull request:
Hi!

I implemented new grabber class for [tim](https://www.sick.com/jp/en/detection-and-ranging-solutions/2d-lidar-sensors/tim5xx/c/g292754) (2d-LiDAR sensor created by SICK).

Please tell me if you have any questions or suggestions!

Following image is demo movie.
I tested it with what I built and installed
![tim_grabber_demo_for_pcl](https://user-images.githubusercontent.com/42496938/94568075-09ac1a80-02a7-11eb-8892-657ae6494340.gif)

It's my first pull request, so I'm sorry if I'm rude.

Thank you for reading :)

kunaltyagi reviewed (changes requested):
Thanks for the contribution. @ysuzuki19. Could you please include tests as well?

Instead of the long-form header, please use the shorter SPDX based header. Grep for SPDX in PCL to find examples.

ysuzuki19 commented:
Thank you for comment. @kunaltyagi !


>Could you please include tests as well?

Sorry, I haven't written an `include test`, is there a manual somewhere?

>Instead of the long-form header, please use the shorter SPDX based header. Grep for SPDX in PCL to find examples.

I am not sure `shorter SPDX based header`, is it different from the description at the bottom of `CONTRIBUTING.md`?
Perhaps a line like following?
 
```cpp
// SPDX-License-Identifier: BSD-3-Clause
```

kunaltyagi commented:
Please check the file `filters/include/pcl/filters/experimental/functor_filter.h` to check how the SPDX include should look like. You're pretty close though.

Regarding the tests, I meant that the PR needs to have tests for your code so that it'll be maintainable for a long time after now. Please check `test/io/test_grabbers.cpp` for older grabber tests. PCL uses GTest framework to make it easier to write and run the tests.

Apologies for not adding relevant hyperlinks.

ysuzuki19 commented:
Thank you for teaching in an easy-to-understand. @kunaltyagi 

I changed to `shorter SPDX based header`. Found it :)


I have an question about TEST...
The class I added assumes LAN communication with the lidar sensor. Isn't it impossible to write a TEST?

I searched TEST for grabber of other sensors (real_sense_grabber, vlp_grabber, etc.) for reference, but could not find it.

The command below returned only {`#include <pcl/io/pcd_grabber.h>`, `#include <pcl/io/image_grabber.h>`}
```sh
/pcl/test/io $  cat * | grep include.*_grabber
```

kunaltyagi commented on io/src/tim_grabber.cpp:
```diff
@@ -0,0 +1,210 @@
+  std::stringstream ss (received_packet_.data ());
+  std::string str;
+
+  for (int i=0; i<26; ++i)
```
Some details on why you're doing this would be nice. A brief overview of the protocol would also be nice

kunaltyagi commented on io/src/tim_grabber.cpp:
```diff
@@ -0,0 +1,210 @@
+void
+pcl::TimGrabber::parseTimPacket ()
+{
+  std::stringstream ss (received_packet_.data ());
```
istringstream?

kunaltyagi commented on io/src/tim_grabber.cpp:
```diff
@@ -0,0 +1,210 @@
+
+  distances_.resize (amount_of_data_);
+  for (auto& distance : distances_) {
+    ss >> str;
```
Wil the following work?
```
ss.setf(std::hex);
for (auto& distance:distances_) {
  ss >> distance;
  distance /= 1000.0;  // convert from [mm] to [m]
}
```

kunaltyagi replied:
Did you try `ss >> std::hex >> distance`? (Also, it has to be `setf` not `self` (T not L))

ysuzuki19 replied:
Sorry, I got a typo when write comment.

Yes, I tried it (and retry it now). But it does not work.

It seems that std :: hex cannot be streamed to float type
So once I put a variable of type ʻint`, it worked.

```
int d;
for (auto& distance : distances_) {
  ss >> std::hex >> d;
  distance = d / 1000.0;
}
```

Thank you!

kunaltyagi commented on io/src/tim_grabber.cpp:
```diff
@@ -0,0 +1,210 @@
+void
+pcl::TimGrabber::updateLookupTables () {
+  //// note: cos, sin are the same size
+  if (cos_dynamic_lookup_table_.size () != amount_of_data_)
```
How frequent is this expected to happen?

kunaltyagi commented on io/src/tim_grabber.cpp:
```diff
@@ -0,0 +1,210 @@
+///////////////////////////////////////////////////////////////////////////////////////////////////
+bool
+pcl::TimGrabber::isValidPacket () const {
+  return received_packet_.data ()[length_-1] == '\03';
```
Use back instead?

kunaltyagi commented on io/src/tim_grabber.cpp:
```diff
@@ -0,0 +1,210 @@
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+void
+pcl::TimGrabber::parseTimPacket ()
```
Separate out into 2 functions: parsePacketHeader and parsePacketBody or something similar based on protocol details

kunaltyagi reviewed (changes requested):
You can't test the network connection, but you can test the parseTimPacket function

ysuzuki19 reviewed (commented):
> You can't test the network connection, but you can test the parseTimPacket function

Year, thank you for teaching.
I understood what you were saying.
And I learned GoogleTest and read `pcl/test/io/test_grabbers.cpp`.

By following URL.
https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#testing-private-code

When applying GTEST to private members, I need `FRIEND_TEST()` in my header.
May I wrote it in `tim_grabber.h` ?

kunaltyagi commented:
> When applying GTEST to private members, I need FRIEND_TEST() in my header

Your test class can simply inherit from the TimGrabber class, giving it access to protected members. If your parsePacket is protected, then the issue will be resolved in an easier manner than adding extra dependency to core libpcl_io

ysuzuki19 commented:
> Your test class can simply inherit from the TimGrabber class, giving it access to protected members. If your parsePacket is protected, then the issue will be resolved in an easier manner than adding extra dependency to core libpcl_io

Thank you teaching!

I understood everything you were saying and was able to resolve it.

I created `/test/io/test_tim_grabber.cpp` :)
I included `updateLookupTable()` and `toPointClouds()` as well as `processTimPackets()`

p.s.
When I use std :: hex, VS2017 Build is Failed, isn't it okay?

kunaltyagi commented on io/src/tim_grabber.cpp:
```diff
@@ -0,0 +1,264 @@
+    ss >> str;
+  //// last value is Amount of data
+
+  amount_of_data_ = std::stoi (str, nullptr, 16);
```
Please skip the initial packets without processing them in loop, eg:
```cpp
  amount_of_data_ = std::stoi (str, <sum of bytes for the 26 fields>, 16);
```

ysuzuki19 replied:
In the packet, It contains values that have nothing to do with the `amount_of_data_`.
The 26th value is the `amount_of_data_`.

More to say, I don't know the bit position until it is separated by spaces.

Moreover, you can't find the beginning of the data without using a loop.


I have another idea, eg:

```cpp
for (short i=0; i<25; ++i) {
  std::string str; // short scope
  ss >> str;
}
ss >> std::hex >> amount_of_data_; // easy to read
```

kunaltyagi commented:
Will check the tests a bit later

kunaltyagi commented on io/include/pcl/io/tim_grabber.h:
```diff
@@ -0,0 +1,123 @@
+namespace pcl
+{
+
+class PCL_EXPORTS TimGrabber : public Grabber
```
Explain the details here. Move the link to the protocol description here as well

kunaltyagi commented on io/src/tim_grabber.cpp:
```diff
@@ -0,0 +1,265 @@
+  //// Amount of data
+  /////////////////////////////////////////////////////
+
+  for (short i=0; i<25; ++i) {
+    std::string str;
+    ss >> str;
+  }
+
+  ss >> std::hex >> amount_of_data_;
```
Reading this isn't needed. You can literally reach in to grab the data and save time (see one of my previous message regarding this)

ysuzuki19 replied:
Year!

I changed function argument type and applied it.
That reduced the number of inputs to the stream.

In addition, it seemed possible, so I rewrote `setf()`.

kunaltyagi commented on test/io/tim_sequences/packets/packet_01.csv:
```diff
@@ -0,0 +1 @@
+sRA LMDscandata 1 1 1291B11 0 0 AED5 AED7 FDB36397 FDB3779F 0 0 1 0 0 5DC A2 0 1 DIST1 3F800000 00000000 FFF92230 D05 32B 6C8 6E5 708 786 7DA 7FA 80C 81C 829 83A 84D 85D 870 889 88F 8AC 8C1 8D0 8E5 8FD 913 92E 546 524 97F 98C 9A3 9BF 9D9 9F1 A0E A2D A3F A3C A36 A33 A32 A20 A1E A18 A0D A05 A02 9FF 9F6 9F2 9F0 9E3 9DD 9D9 9D1 9CE 9C5 9BF 9BC 9B6 9AE 9AE 9A2 9A2 999 99B 998 98C 98C 990 980 983 980 9 ...
```
This is not a csv

kunaltyagi commented on test/io/test_tim_grabber.cpp:
```diff
@@ -0,0 +1,121 @@
+  public: using TimGrabber::toPointClouds;
+};
+
+TEST (PCL, TestableTimGrabber)
```
Please add an expected test:
* Create a `SetUp` where you create 2 things:
  * a point cloud (say a line at 45 degree from -135 to 135 degree)
  * a string with those distance values in hex (basically the format)
* Check that the 2 are equivalent

This will test that the code adheres to some "API". Your code requires a ground truth pcd to be present (and if you created the pcd using your code, we'd be testing your implementation, not the actual protocol)

larshg commented:
Hi @ysuzuki19 

Please take a look [here](https://github.com/google/googletest/blob/master/googletest/docs/primer.md) on how to properly implement the Setup of unit test using the google test framework.

Thanks for your contribution!

ysuzuki19 commented:
Hi @larshg .

oh, google test framework seems too complicated and difficult for me. 😂

But I was able to implement by referring to the links you sent me.

Thanks for your comment!

kunaltyagi commented on test/io/test_tim_grabber.cpp:
```diff
@@ -0,0 +1,141 @@
+
+class TestableTimGrabber : public pcl::TimGrabber {
+  using TimGrabber::TimGrabber;
+  public: using TimGrabber::point_cloud_xyz_ptr_;
```
Using one public is enough. (Hint: use struct which makes all the public redundant)

ysuzuki19 replied:
Done

kunaltyagi commented on test/io/test_tim_grabber.cpp:
```diff
@@ -0,0 +1,141 @@
+    virtual void TearDown () {
+    }
+
+    static std::vector<std::string>* packets_;
```
No need of static or pointer. Simple stack variables will suffice

ysuzuki19 replied:
Done

kunaltyagi commented on test/io/test_tim_grabber.cpp:
```diff
@@ -0,0 +1,141 @@
+    {
+      const PointT &correct_point = correct_clouds_->at(i).at(j);
+      const PointT &answer_point = answer_cloud->at(j);
+      if (std::isnan (correct_point.x))
```
Insert Nan in the data in order to actually test the nan checks

ysuzuki19 replied:
I wrote this check (`isnan()`) referenced by `test/io/test_grabbers.cpp`.

But  I think about it again, these comparison (some `isnan()`) are not necessary.
Because in this test, `0<= x,y,z <= 20` was generated by random_generator.

Required comparison are only
- correct.x == answer.x
- correct.y == answer.y
- correct.z == answer.z

kunaltyagi replied:
Is there any situation in which nan will be in the input from the real sensor? If no, then all these checks are not needed in the grabber (and the test)

ysuzuki19 replied:
No!

I used Tim to send light to the night sky now,  but there was no NAN value.

※ The maximum distance of Tim is 20m.

larshg commented on test/io/test_tim_grabber.cpp:
```diff
@@ -0,0 +1,106 @@
+    static void TearDownTestCase () {
+    }
+
+    virtual void SetUp () {
```
```suggestion
    virtual void SetUp() override {
```

larshg replied:
some more places missing the override keyword.

ysuzuki19 replied:
Thanks.

may I use `SetUp () override` instead of `SetUp() override`?

And removed unused functions!

larshg replied:
I don't mind, but I think if you run the make format (not entirely sure how that part works :P ) the extra parentheses will get removed anyway. Noticed it when reviewing another PR which just does that (correcting format).

ysuzuki19 commented:
Fixed a bug in initialization

It was divided by 0 of float and bad_alloc

```cpp
const float angle_step = angle_range_ / static_cast<float>(amount_of_data_);
`

kunaltyagi commented on test/io/test_tim_grabber.cpp:
```diff
@@ -0,0 +1,97 @@
+using PointT = pcl::PointXYZ;
+using CloudT = pcl::PointCloud<PointT>;
+
+class TestableTimGrabber : public pcl::TimGrabber {
```
struct since everything is public

kunaltyagi commented on test/io/test_tim_grabber.cpp:
```diff
@@ -0,0 +1,97 @@
+		using TimGrabber::toPointClouds;
+};
+
+std::vector<std::string> packets_;
+std::vector<CloudT> correct_clouds_;
+TestableTimGrabber grabber_;
```
Is there any benefit to these variables being global vs them being members of TimGrabberTest class?

kunaltyagi reviewed (changes requested):
Please remove the 10 MB binary file test_tim_grabber and clean up the history (so this isn't included in PCL's git repo).

Apart from that, it mostly looks good to me 😄

kunaltyagi reviewed (commented):
@ysuzuki19 Remove and rebase the binary file please.

ysuzuki19 commented:
Oops!
A built binary was added
Thanks for your mention.

I will try rebase but wait a minute please ...

ysuzuki19 commented:
Rebased!😊

kunaltyagi commented:
Could you squash you commits so they make more sense? Eg: add the code and tests in separate commit. Else we'll have to squash the entire history (too many weird commit messages otherwise)

kunaltyagi commented on test/io/test_tim_grabber.cpp:
```diff
@@ -0,0 +1,96 @@
+#include <pcl/test/gtest.h>
```
Just noticed: missing SPDX header

ysuzuki19 commented:
I finally understood the meaning of squash.😄
And solve it!

ysuzuki19 commented:
I resolved conflict in `test/io/CMakeLists.txt`.

ysuzuki19 commented:
@kunaltyagi 

Now that the conflict has been resolved, could I re-review it?

If it looks okay, i do rebase.

kunaltyagi commented:
Rebase requires using the cli.

You can update the master branch to latest, then do:
```
git rebase -i master
```

If there's a conflict, please resolve it, add it, (maybe commit it) and continue the rebase. The cli is quite informative

larshg commented on io/include/pcl/io/tim_grabber.h:
```diff
@@ -0,0 +1,166 @@
+#include <pcl/point_types.h>
+#include <pcl/point_cloud.h>
+#include <boost/asio.hpp>
+#include <boost/thread.hpp>
```
Is this required? or is it all std::thread?

larshg commented on io/include/pcl/io/tim_grabber.h:
```diff
@@ -0,0 +1,166 @@
+    unsigned int wait_time_milliseconds_ = 0;
+
+    pcl::EventFrequency frequency_;
+    mutable boost::mutex frequency_mutex_;
```
Ahh, here is boost::thread used? Can we maybe use std::mutex?

larshg commented on io/src/tim_grabber.cpp:
```diff
@@ -0,0 +1,238 @@
+float
+pcl::TimGrabber::getFramesPerSecond () const
+{
+  boost::mutex::scoped_lock lock( frequency_mutex_ );
```
std::scoped_lock?

ysuzuki19 commented:
@kunaltyagi 
> If there's a conflict, please resolve it, add it, (maybe commit it) and continue the rebase. The cli is quite informative

Thank you for teaching git for me.

I rebase master and resolve conflict and push now.

ysuzuki19 commented:
@kunaltyagi

sorry!
commit id was changed by force push from git cli for rebase, could you re-approve it?

kunaltyagi commented:
There are comments by larshg. Could you address them before I re-review?

larshg commented:
@ysuzuki19 is it possible to change the boost mutexes to std::mutexes and reduce the required boost includes?
Please see the comments I made.

ysuzuki19 commented:
@larshg 
Do you think about using `std::lock_guard` instead of `boost::scoped_lock()`?

I actually want to use `std::scoped_lock`  than `std::lock_guard`.
But PCL require C++14 and `std::scoped_lock()` is [since C++17](https://en.cppreference.com/w/cpp/thread/scoped_lock) 😭

so i use `boost::scoped_lock()` and `boost::mutex`.

In the future, when PCL require C++17, this code are rewritable to `std::scoped_lock()` and `std::mutex`.

Thanks!

larshg commented:
Ah, makes sense. Must have missed the c++17 requirement for scoped_lock!

ysuzuki19 commented:
@larshg 

Thank you for taking the time and approved commit 😃

kunaltyagi commented on io/src/tim_grabber.cpp:
```diff
@@ -0,0 +1,238 @@
+
+  //// packet contains header and body.
+  //// header has 26 spaces, the rest is body
+  std::string header (packet.substr (0, header_body_splitter));
+  std::string body   (packet.substr (header_body_splitter + 1));
```
Thoughts on using string_view here?

It's in Boost since 1.62 at least (18.04 LTS is more modern than that)

kunaltyagi replied:
Thanks for experimenting.

kunaltyagi reviewed (commented):
Just minor clarification and comments.

Sorry for being so late on this

ysuzuki19 commented:
@kunaltyagi 

Thank you for review.
I replied comment.

kunaltyagi commented:
I don't see any comments @ysuzuki19

ysuzuki19 commented:
Sorry, I've done.
I didn't understand `Pending` statement in review.

@kunaltyagi

kunaltyagi commented on io/src/tim_grabber.cpp:
```diff
@@ -0,0 +1,238 @@
+
+void
+pcl::TimGrabber::parsePacketBody (std::string const& body) {
+  std::istringstream ss (body);
```
Let's make it class variable in that case

kunaltyagi commented on io/src/tim_grabber.cpp:
```diff
@@ -0,0 +1,238 @@
+  if (!isValidPacket ()) {
+    //// Countup wait_time
+    //// Because fit wait_time to network latency
+    wait_time_milliseconds_++;
```
Maybe a future Todo for improving this behavior is better.

ysuzuki19 commented:
Sorry I made mistakes. @kunaltyagi 

I noticed it yesterday, found out the issue and fixed it! This was caused by `istringstream` is not cleared.
With this commit, all CI have passed.
